### ISSUE TITLE
پیاده سازی مقاله اول

### ISSUE THREAD

**Issue by romosij200218-prog** [Amir13849271741537/Machine-learning-#32]
M.h:
############################################################
#  پیاده‌سازی کامل یک مدل CNN روی دیتاست CIFAR-10
#  شامل: آموزش، ارزیابی، تحلیل خطا و محاسبه معیارهای عملکرد
#  نویسنده: ChatGPT
############################################################

import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from sklearn.metrics import precision_score, recall_score, f1_score
import matplotlib.pyplot as plt
import numpy as np
import time

############################################################
# 1. بارگذاری دیتاست CIFAR-10
#    این دیتاست شامل 10 کلاس و 60000 تصویر است.
############################################################

transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5,), (0.5,))
])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=64,
                                          shuffle=True)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=64,
                                         shuffle=False)

classes = trainset.classes

############################################################
# 2. تعریف یک مدل ساده CNN
#    (برای شبیه‌سازی مقایسه‌ با مدل‌های مقاله)
############################################################

class SimpleCNN(nn.Module):
    def init(self):
        super(SimpleCNN, self).init()
        self.net = nn.Sequential(
            nn.Conv2d(3, 32, 3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2),

            nn.Conv2d(32, 64, 3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2),

            nn.Flatten(),
            nn.Linear(64 * 8 * 8, 256),
            nn.ReLU(),
            nn.Linear(256, 10)
        )

    def forward(self, x):
        return self.net(x)

model = SimpleCNN()
device = "cuda" if torch.cuda.is_available() else "cpu"
model.to(device)

############################################################
# 3. تعریف تابع هزینه و بهینه‌ساز
############################################################

criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

############################################################
# 4. آموزش مدل + اندازه‌گیری زمان آموزش (Training Time)
############################################################

start_time = time.time()

epochs = 5
for epoch in range(epochs):
    running_loss = 0
    for images, labels in trainloader:
        images, labels = images.to(device), labels.to(device)

        optimizer.zero_grad()
        outputs = model(images)

        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        running_loss += loss.item()

    print(f"Epoch [{epoch+1}/{epochs}] Loss: {running_loss/len(trainloader)}")

training_time = time.time() - start_time
print(f"\n⏱ زمان آموزش: {training_time:.2f} ثانیه\n")

############################################################
# 5. ارزیابی مدل + محاسبه Accuracy ،Precision ،Recall ،F1
############################################################

y_true = []
y_pred = []

model.eval()
with torch.no_grad():
    for images, labels in testloader:
        images, labels = images.to(device), labels.to(device)

        outputs = model(images)
        _, predicted = torch.max(outputs, 1)

        y_true.extend(labels.cpu().numpy())
        y_pred.extend(predicted.cpu().numpy())

accuracy = np.mean(np.array(y_true) == np.array(y_pred))
precision = precision_score(y_true, y_pred, average="macro")
recall = recall_score(y_true, y_pred, average="macro")
f1 = f1_score(y_true, y_pred, average="macro")

print("📊 نتایج ارزیابی مدل:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

############################################################
# 6. اندازه‌گیری زمان پیش‌بینی یک تصویر (Inference Time)
############################################################

test_image = testset[0][0].unsqueeze(0).to(device)

start_time = time.time()
with torch.no_grad():
    _ = model(test_image)
inference_time = time.time() - start_time

print(f"\n⚡ زمان پیش‌بینی یک تصویر: {inference_time*1000:.4f} میلی‌ثانیه\n")

############################################################
# 7. تحلیل خطا (Error Analysis)
#    نمایش 10 تصویر که مدل اشتباه پیش‌بینی کرده است
############################################################

errors_img = []
errors_true = []
errors_pred = []

count = 0
for images, labels in testloader:
    images, labels = images.to(device), labels.to(device)
    outputs = model(images)
    _, predicted = torch.max(outputs, 1)

    for i in range(len(labels)):
        if predicted[i] != labels[i] and count < 10:
            errors_img.append(images[i].cpu())
            errors_true.append(labels[i].item())
            errors_pred.append(predicted[i].item())
            count += 1

    if count >= 10:
        break

# نمایش تصاویر اشتباه
plt.figure(figsize=(12, 5))
for idx in range(10):
    plt.subplot(2, 5, idx+1)
    img = errors_img[idx] / 2 + 0.5  # برگرداندن Normalize
    plt.imshow(np.transpose(img.numpy(), (1, 2, 0)))
    plt.title(f"واقعی: {classes[errors_true[idx]]}\nپیش‌بینی: {classes[errors_pred[idx]]}")
    plt.axis("off")

plt.show()

############################################################
# پایان کد 
############################################################
